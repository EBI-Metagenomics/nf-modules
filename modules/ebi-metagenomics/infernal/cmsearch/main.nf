
process INFERNAL_CMSEARCH {

    tag "$meta.id"

    label 'process_low'

    conda "bioconda::infernal=1.1.5"
    container "${ workflow.containerEngine == 'singularity' && !task.ext.singularity_pull_docker_container ?
        'https://depot.galaxyproject.org/singularity/infernal:1.1.5--pl5321h031d066_1':
        'biocontainers/infernal:1.1.5--pl5321h031d066_1' }"

    input:
    tuple val(meta), path(seqdb)
    path covariance_model_database

    output:
    tuple val(meta), path("*.cmsearch_matches.tbl.gz"), emit: cmsearch_tbl
    path "versions.yml"                            , emit: versions

    when:
    task.ext.when == null || task.ext.when

    script:
    def args = task.ext.args ?: ''
    def prefix = task.ext.prefix ?: "${meta.id}"

    def is_compressed = seqdb.name.endsWith(".gz")
    def seqdb_name = seqdb.name.replace(".gz", "")
    """
    if [ "$is_compressed" == "true" ]; then
        gzip -c -d $seqdb > $seqdb_name
    fi

    cmsearch \\
        --cpu $task.cpus \\
        $args \\
        --tblout ${prefix}.cmsearch_matches.tbl \\
        $covariance_model_database/*.cm \\
        $seqdb_name

    gzip -n ${prefix}.cmsearch_matches.tbl

    cat <<-END_VERSIONS > versions.yml
    "${task.process}":
        cmsearch: \$(cmsearch -h | grep -o '^# INFERNAL [0-9.]*' | sed 's/^# INFERNAL *//')
    END_VERSIONS
    """

    stub:
    def args = task.ext.args ?: ''
    def prefix = task.ext.prefix ?: "${meta.id}"
    """
    touch ${prefix}.cmsearch_matches.tbl
    gzip ${prefix}.cmsearch_matches.tbl

    cat <<-END_VERSIONS > versions.yml
    "${task.process}":
        cmsearch: \$(cmsearch -h | grep -o '^# INFERNAL [0-9.]*' | sed 's/^# INFERNAL *//')
    END_VERSIONS
    """
}
