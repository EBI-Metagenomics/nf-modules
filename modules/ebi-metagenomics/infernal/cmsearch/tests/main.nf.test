nextflow_process {

    name "Test Process INERNAL/CMSEARCH"
    script "../main.nf"
    process "INFERNAL_CMSEARCH"
    config "./nextflow.config"

    tag "modules"
    tag "modules_ebimetagenomicsgit"
    tag "infernal"
    tag "cmsearch"
    tag "infernal/cmsearch"

    test("infernal cmsearch") {
        when {
            process {
                """
                input[0] = [
                    [ id:'test', single_end:false ],                              // meta map
                    file("${moduleDir}/tests/example.fa.gz", checkIfExists: true) // seqdb
                    ]
                input[1] = [file("${moduleDir}/tests/rfam", checkIfExists: true)]
                """
            }
        }

        then {
            assertAll(
                { assert process.success },
                { assert snapshot(path(process.out.cmsearch_tbl.get(0).get(1)).contains("# [ok]")) },
                { assert snapshot(process.out.versions).match('versions') }
            )
        }
    }

    test("stub") {

        options "-stub"

        when {
            process {
                """
                input[0] = [
                    [ id:'test', single_end:false ], // meta map
                    file("${moduleDir}/tests/example.fa.gz", checkIfExists: true)
                    ]
                input[1] = file("${moduleDir}/tests/rfam", checkIfExists: true)
                """
            }
        }

        then {
            assertAll(
                { assert process.success },
                { assert snapshot(process.out).match() }
            )
        }

    }

}
