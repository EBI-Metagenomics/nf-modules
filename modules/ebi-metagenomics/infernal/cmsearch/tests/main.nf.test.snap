{
    "versions": {
        "content": [
            [
                "versions.yml:md5,643c3b5b34f868d87511259bc33d3f3a"
            ]
        ],
        "meta": {
            "nf-test": "0.9.2",
            "nextflow": "24.10.5"
        },
        "timestamp": "2025-03-19T15:36:01.996661"
    },
    "stub": {
        "content": [
            {
                "0": [
                    [
                        {
                            "id": "test",
                            "single_end": false
                        },
                        "test.cmsearch_matches.tbl.gz:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "1": [
                    "versions.yml:md5,643c3b5b34f868d87511259bc33d3f3a"
                ],
                "cmsearch_tbl": [
                    [
                        {
                            "id": "test",
                            "single_end": false
                        },
                        "test.cmsearch_matches.tbl.gz:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "versions": [
                    "versions.yml:md5,643c3b5b34f868d87511259bc33d3f3a"
                ]
            }
        ],
        "meta": {
            "nf-test": "0.9.2",
            "nextflow": "24.10.5"
        },
        "timestamp": "2025-03-21T18:50:44.728532"
    }
}